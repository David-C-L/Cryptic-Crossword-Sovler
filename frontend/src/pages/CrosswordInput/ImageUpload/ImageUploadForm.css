.UploadSection {
    text-align: center;
}

.UploadLabel {
    margin-bottom: 10px;
}