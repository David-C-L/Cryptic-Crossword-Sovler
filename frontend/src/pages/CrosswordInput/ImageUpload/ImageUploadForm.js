import "./ImageUploadForm.css"
import { useCallback, useEffect, useState } from 'react';
import axios from 'axios';

function ImageUploadForm(props) {

    // Constants to store image files
    const [grid, setGrid] = useState(null);
    const [across, setAcross] = useState(null);
    const [down, setDown] = useState(null);

    // Information about dimensions of grid
    const [noOfRows, setNoOfRows] = useState(0);
    const [noOfColumns, setNoOfColumns] = useState(0);

    // Boolean flag for showing error message if answer can not be found
    const [showError, setShowError] = useState(false);

    // const baseURL = (Number(process.env.REACT_APP_PRODUCTION) === 1 ? process.env.REACT_APP_EXTERNAL : process.env.REACT_APP_INTERNAL)
    const baseURL = "http://0.0.0.0:5000"
    const solveEntireGridURL = baseURL + '/uploadImage';
    const eventSource = baseURL + '/stream';

    const eL = useCallback((e) => {
        window.location = "/crosswordSolver"
    }, [])

    useEffect(() => {
        const source = new EventSource(eventSource);
        source.addEventListener('change_screen', eL)
        return () => source.removeEventListener("change_screen", eL)
    }, [eL, eventSource])

    const handlePuzzleToSolve = (e) => {
        e.preventDefault();
        setGrid(e.target.files[0]);
    }

    const handleAcrossClues = (e) => {
        e.preventDefault();
        setAcross(e.target.files[0]);
    }

    const handleDownClues = (e) => {
        e.preventDefault();
        setDown(e.target.files[0]);
    }

    const handleNoOfRowsChange = (e) => {
        e.preventDefault();
        setNoOfRows(e.target.value);
    };

    const handleNoOfColumnsChange = (e) => {
        e.preventDefault();
        setNoOfColumns(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();
        props.setResult(false);

        // Wrapping input metadata into Form data object
        let formData = new FormData()
        formData.append('noOfColumns', noOfColumns);
        formData.append('noOfRows', noOfRows);
        formData.append('grid', grid);
        formData.append('across', across);
        formData.append('down', down);

        axios.post(solveEntireGridURL, formData, {
            'Content-Type': 'multipart/form-data'
        }).then(response => {
            props.setValueGrid(response.data.grid);
            props.setResult(true)
        }).catch(() => {
            setShowError(true)
        })
    }

    return (
        <form>

            <div className={"row mb-5"}>
                <div className="col form-group">
                    <label htmlFor="puzzleUpload">Crossword Puzzle</label>
                    <input className="form-control" id="puzzleUpload" type="file" accept={".jpg,.png"} onChange={handlePuzzleToSolve}/>
                </div>
            </div>

            <div className={"row mb-5"}>
                <div className="col form-group">
                    <label htmlFor="acrossCluesUpload">Across Clues</label>
                    <input className="form-control" id="acrossCluesUpload" type="file" accept={".jpg,.png"} onChange={handleAcrossClues}/>
                </div>
            </div>

            <div className={"row mb-5"}>
                <div className="col form-group">
                    <label htmlFor="downCluesUpload">Down Clues</label>
                    <input className="form-control" id="downCluesUpload" type="file" accept={".jpg,.png"} onChange={handleDownClues}/>
                </div>
            </div>

            <div className={"row mb-5"}>
                <div className="col form-group">
                    <label htmlFor="inputNoOfRows">Number of Rows</label>
                    <input type="number" className="form-control" id="inputNoOfRows" aria-describedby="noOfRows" value={noOfRows || ''} onChange={handleNoOfRowsChange} />
                </div>
                <div className="col form-group">
                    <label htmlFor="inputNoOfColumns">Number of Columns</label>
                    <input type="number" className="form-control" id="inputNoOfColumns" aria-describedby="noOfColumns" value={noOfColumns || ''} onChange={handleNoOfColumnsChange} />
                </div>
            </div>

            <div className={"row mb-5"}>
                <div className={"col text-center"}>
                    <button type="submit" className="my-3 btn btn-primary" onClick={handleSubmit}>Submit</button>
                    {
                        showError &&
                        <div className="alert alert-warning text-center">
                            <strong>Oh no!</strong> No answer nor explanation could be found. Please check your input.
                        </div>
                    }
                </div>
            </div>
        </form>
    )
}

export default ImageUploadForm;