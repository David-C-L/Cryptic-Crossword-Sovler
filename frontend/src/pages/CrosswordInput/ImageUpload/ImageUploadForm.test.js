import React from "react";
import { unmountComponentAtNode } from "react-dom";
import renderer from 'react-test-renderer';
import ImageUploadForm from './ImageUploadForm'

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('renders correctly', () => {
  const tree = renderer
    .create(<ImageUploadForm />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});