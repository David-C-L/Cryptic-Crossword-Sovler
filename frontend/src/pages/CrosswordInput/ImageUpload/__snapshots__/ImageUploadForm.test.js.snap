// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form>
  <div
    className="row mb-5"
  >
    <div
      className="col form-group"
    >
      <label
        htmlFor="puzzleUpload"
      >
        Crossword Puzzle
      </label>
      <input
        accept=".jpg,.png"
        className="form-control"
        id="puzzleUpload"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
  <div
    className="row mb-5"
  >
    <div
      className="col form-group"
    >
      <label
        htmlFor="acrossCluesUpload"
      >
        Across Clues
      </label>
      <input
        accept=".jpg,.png"
        className="form-control"
        id="acrossCluesUpload"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
  <div
    className="row mb-5"
  >
    <div
      className="col form-group"
    >
      <label
        htmlFor="downCluesUpload"
      >
        Down Clues
      </label>
      <input
        accept=".jpg,.png"
        className="form-control"
        id="downCluesUpload"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
  <div
    className="row mb-5"
  >
    <div
      className="col form-group"
    >
      <label
        htmlFor="inputNoOfRows"
      >
        Number of Rows
      </label>
      <input
        aria-describedby="noOfRows"
        className="form-control"
        id="inputNoOfRows"
        onChange={[Function]}
        type="number"
        value=""
      />
    </div>
    <div
      className="col form-group"
    >
      <label
        htmlFor="inputNoOfColumns"
      >
        Number of Columns
      </label>
      <input
        aria-describedby="noOfColumns"
        className="form-control"
        id="inputNoOfColumns"
        onChange={[Function]}
        type="number"
        value=""
      />
    </div>
  </div>
  <div
    className="row mb-5"
  >
    <div
      className="col text-center"
    >
      <button
        className="my-3 btn btn-primary"
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </div>
</form>
`;
